exporter/exporterparser/zipkin.go

```diff
@@ -66,22 +95,99 @@ func ZipkinExportersFromYAML(config []byte) (tes []exporter.TraceExporter, doneF
 	if zc.LocalEndpointURI != "" {
 		localEndpointURI = zc.LocalEndpointURI
 	}
-	// TODO(jbd): Propagate hostport and more metadata from each node.
+	endpoint := zc.EndpointURL()
+	zle, err := newZipkinExporter(endpoint, serviceName, localEndpointURI)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Cannot configure Zipkin exporter: %v", err)
+	}
+	tes = append(tes, zle)
+	doneFns = append(doneFns, zle.stop)
+	return
+}
+
+func newZipkinExporter(finalEndpointURI, defaultServiceName, defaultLocalEndpointURI string) (*zipkinExporter, error) {
+	localEndpoint, err := openzipkin.NewEndpoint(defaultServiceName, defaultLocalEndpointURI)
+	if err != nil {
+		return nil, err
+	}
+	reporter := http.NewReporter(finalEndpointURI)
+	defaultExporter := zipkin.NewExporter(reporter, localEndpoint)
+	zle := &zipkinExporter{
+		endpointURI:             finalEndpointURI,
+		defaultExporter:         defaultExporter,
+		defaultServiceName:      defaultServiceName,
+		defaultLocalEndpointURI: defaultLocalEndpointURI,
+		serviceNameToExporter:   make(map[string]*zipkin.Exporter),
+	}
+
+	// Ensure that we add the default reporter's Close functions
+	zle.shutdownFns = append(zle.shutdownFns, reporter.Close)
+
+	return zle, nil
+}
+
+// exporterForNode firstly tries to find a memoize OpenCensus-Go Zipkin exporter
+// appropriate for this node. If it doesn't find any, it returns the default exporter
+// that was created at start-time.
+func (ze *zipkinExporter) exporterForNode(node *commonpb.Node) *zipkin.Exporter {
+	ze.mu.Lock()
```

Comment:
The initial checks don't seem to need be under a writer lock. From what I see it is intended to protect ze.serviceNameToExporter, any reason to not attemp a read lock first and only after that failing trying a writer lock (in the minimal scope possible)?

Reply:
>The initial checks don't seem to need be under a writer lock.

This call is only used dynamically during every call to trace.ExportSpan while newZipkinExporter is created once at startTime and also the shared entity hasn't yet been created. That's why we need a lock here and not there. For the other checks we also have locks.

>From what I see it is intended to protect ze.serviceNameToExporter,

It also protects `.stop`
<img width="377" alt="screen shot 2018-10-25 at 11 13 50 am" src="https://user-images.githubusercontent.com/4898263/47521153-12884c00-d847-11e8-8fd6-6c6caf853e0a.png">
and is `serviceNameToExporter` is used by `ExportSpanData` so all its methods are protected by the same lock.

>any reason to not attemp a read lock first and only after that failing trying a writer lock (in the minimal scope possible)?

Great question: a readLock then a write lock are going to complicate the logic at the expense of readability and perhaps add more code because an error will need read unlocking on a separate error message. I think if we deem this to be in a very hot path, then we can apply that.

Reply:
Thanks @odeke-em - it seems reasonable: contention is expected to be low and we can improve if we detect the need when stress testing.

One small note: since `mu` just uses Lock()/Unlock() it will be clear if it is a `sync.Mutex` instead of a `sync.RWMutex`.

Reply:
@pjanotti, gotcha thanks I'll update it to `sync.Mutex` instead of `sync.RWMutex`.

Reply:
And done, thanks @pjanotti